app/org/sagebionetworks/bridge/hibernate/AccountPersistenceExceptionConverter.java

```diff
@@ -0,0 +1,79 @@
+import org.sagebionetworks.bridge.models.accounts.AccountId;
+import org.springframework.stereotype.Component;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+
+@Component
+public class AccountPersistenceExceptionConverter implements PersistenceExceptionConverter {
+    private final AccountDao accountDao;
+    
+    public AccountPersistenceExceptionConverter(AccountDao accountDao) {
+        this.accountDao = accountDao;
+    }
+    
+    @Override
+    public RuntimeException convert(PersistenceException exception, Object entity) {
+        // The sequence of type-checking and unwrapping of this exception is significant as unfortunately, 
+        // the hierarchy of wrapped exceptions is very specific. 
+        if (exception instanceof OptimisticLockException) {
+            return new ConcurrentModificationException(
+                    "Account has the wrong version number; it may have been saved in the background.");
+        }
+        if (exception.getCause() instanceof org.hibernate.exception.ConstraintViolationException) {
+            // The specific error message is buried in the root MySQLIntegrityConstraintViolationException
+            Throwable cause = Throwables.getRootCause(exception);
+            String message = cause.getMessage();
+            if (message != null && entity != null) {
+                HibernateAccount account = (HibernateAccount)entity;
+                // These are the constraint violation messages. To ensure we don't log credentials, we look
+                // up the existing account and its userId to report in the EntityAlreadyExistsException.
+                // Messages:
+                // "Duplicate entry 'api-email@email.com' for key 'Accounts-StudyId-Email-Index'"
+                // "Duplicate entry 'api-+12064953700' for key 'Accounts-StudyId-Phone-Index'"
+                // "Duplicate entry 'api-ext' for key 'Accounts-StudyId-ExternalId-Index'" 
+                EntityAlreadyExistsException eae = null;
+                if (message.matches("Duplicate entry.*for key 'Accounts-StudyId-ExternalId-Index'")) {
+                    eae = createEntityAlreadyExistsException("External ID",
+                            AccountId.forExternalId(account.getStudyId(), account.getExternalId()));
+                } else if (message.matches("Duplicate entry.*for key 'Accounts-StudyId-Email-Index'")) {
+                    eae = createEntityAlreadyExistsException("Email address",
+                            AccountId.forEmail(account.getStudyId(), account.getEmail()));
+                } else if (message.matches("Duplicate entry.*for key 'Accounts-StudyId-Phone-Index'")) {
+                    eae = createEntityAlreadyExistsException("Phone number",
+                            AccountId.forPhone(account.getStudyId(), account.getPhone()));
+                }
+                if (eae != null) {
+                    return eae;
+                }
```

Comment:
The original exception may include the email address or phone number, which may get logged. I recommend converting it to a different exception with a generic message instead.

Reply:
Constraint exceptions now have a generic error message.